cassdegrees/api/degreeRules.py

```diff
@@ -1,52 +1,161 @@
+    that can be stored in the database under [degree].[rules].
     """
-    def buildExpression(self, expressionList):
-        tree = {}
+
+    def __init__(self, expression_list):
+        self.expressionTree = []
+        self.build_expression(expression_list)
+
+    def __init__(self):
+        self.expressionTree = []
+
+    def build_expression(self, expression_list):
+        """
+        Accepts a list of tokens that represent a boolean expression of degree rules
+        and converts them into a k-ary tree of lists and dictionaries that can be
+        stored in JSON format.
+
+        :param expression_list: (list) a list of tokens in the format of a boolean expression.
+                                A token must be one of:
+                                    "(", ")", "AND", "OR", rule
+                                where rule is some dictionary representing a rule, returned from a function below
+        :return: (list) a k-ary tree with rules as leaf nodes, and boolean operators as the other nodes
+        """
+
+        tree = [None]
         stack = []
         node = tree
-        for token in expressionList:
+        for token in expression_list:
             if token == "(":
-                node['left'] = {}
+                node.append([None])
                 stack.append(node)
-                node = node['left']
+                node = node[-1]
             elif token == ")":
                 if len(stack) > 0:
                     node = stack.pop()
                 else:
-                    print("h")
-                    parent = {}
-                    parent["left"] = node
+                    parent = [None]
+                    parent.append(node)
                     node = parent
                     tree = parent
             elif token in ["AND", "OR"]:
-                node['val'] = token
-                node['right'] = {}
+                node[0] = token
+                node.append([None])
                 stack.append(node)
-                node = node['right']
+                node = node[-1]
             else:
-                node['rule'] = token
-                print(stack)
+                node[0] = token
                 parent = stack.pop()
-                print(stack)
                 node = parent
+
+        self.expressionTree = tree
+
         return tree
-        #create rules, return rules. Then take a list of rules in a bool alg expression and build the json
 
-    # Add rule as another AND?
+    def append_rule(self, rule):
+        """
+        Takes a rule that should be independent of the existing rules for a degree
+        and appends it to the degree rules with an AND operator.
+
+        :param rule: (Dict) the rule to be appended
+        """
+
+        if self.expressionTree[0] == "AND":
+            self.expressionTree.append(rule)
+        else:
+            self.expressionTree = ["AND", self.expressionTree, rule]
+
+    def get_rules_as_json(self):
```

Comment:
When generating rules for myself, the first rule was always `null`. Doesn't look intentional

Reply:
Please elaborate? What were you doing to get a null?

Reply:
@aidengeorge97 This is what I did to get `null`:
```
>>> from api.degreeRules import DegreeRules
>>> d_rules = DegreeRules()
>>> d_rules.build_expression([DegreeRules.rule_electives(24)])
>>> print(d_rules.get_rules_as_json())
[null, [{"type": "electives", "units": 24}]]
```

The same thing also happens with more complex rules such as `d_rules.build_expression(['(', DegreeRules.rule_electives(24), 'AND', DegreeRules.rule_electives(16), ')'])`

Reply:
Ahhhhhh okay I didn't account for a scenario where a degree would only have one rule. Cause no degrees have one rule. But, valid, I'll just make it so one rule expressions default to an AND so it's a single term AND (i.e. just the one rule).

Reply:
Okay I solved the issue. Because you had brackets around those terms, it was expecting another set of rules, hence making a null parent node. I've added a function to flatten the tree, so those kinds of null nodes will be removed and their children get moved up. The logic was still right, but this will neaten the trees up and account for a wider range of poorly designed expressions. 

I also noticed an "issue" with append_rule while changing that, which is now fixed. Rather than just appending singular rules, you can now append a full expression list.